File: FIGLet.TS/src/FIGFont.ts
import { SmushingRules } from './SmushingRules.js';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';
import { promises as fs } from 'fs';
import { inflateRaw } from 'zlib';
import { promisify } from 'util';

const inflateRawAsync = promisify(inflateRaw);

/*
 *   ___ ___ ___ ___        _   
 *  | __|_ _/ __| __|__ _ _| |_ 
 *  | _| | | (_ | _/ _ \ ' \  _|
 *  |_| |___\___|_|\___/_||_\__|
 *                              
 */
/**
 * Represents a FIGfont used for rendering text in FIGLet style.
 */
export class FIGFont {
    private static _default: FIGFont | null = null;

    /**
     * Gets the default FIGfont.
     */
    public static async getDefault(): Promise<FIGFont> {
        if (!FIGFont._default) {
            FIGFont._default = await FIGFont.loadDefaultFont();
        }
        if (!FIGFont._default) {
            throw new Error("Default FIGfont not found");
        }
        return FIGFont._default;
    }

    /**
     * Gets the signature of the FIGfont.
     */
    public signature: string = "flf2a";

    /**
     * Gets the hard blank character used in the FIGfont.
     */
    public hardBlank: string = "#";

    /**
     * Gets the height of the FIGfont characters.
     */
    public height: number = 0;

    /**
     * Gets the baseline of the FIGfont characters.
     */
    public baseline: number = 0;

    /**
     * Gets the maximum length of the FIGfont characters.
     */
    public maxLength: number = 0;

    /**
     * Gets the old layout mode of the FIGfont.
     */
    public oldLayout: number = 0;

    /**
     * Gets the print direction of the FIGfont.
     */
    public printDirection: number = 0;

    /**
     * Gets the full layout mode of the FIGfont.
     */
    public fullLayout: number = 0;

    /**
     * Gets the dictionary of characters in the FIGfont.
     */
    public characters: Map<string, string[]> = new Map();

    /**
     * Gets the smushing rules for the FIGfont.
     */
    public smushingRules: SmushingRules = SmushingRules.None;

    /**
     * Gets the comments for the FIGfont.
     */
    public comments: string = "";

    /**
     * Creates a FIGFont from a file.
     * Automatically detects ZIP archives by checking for the PK magic bytes
     * and extracts the first entry, matching the C# FIGFontStream behaviour.
     */
    public static async fromFile(path: string | null): Promise<FIGFont | null> {
        if (!path) {
            return null;
        }

        try {
            const buffer = await fs.readFile(path);

            // Detect ZIP by magic bytes 'PK' (0x50 0x4B)
            if (buffer.length >= 2 && buffer[0] === 0x50 && buffer[1] === 0x4B) {
                const text = await FIGFont.extractFirstZipEntry(buffer);
                return FIGFont.fromText(text);
            }

            return FIGFont.fromText(buffer.toString('utf-8'));
        } catch {
            return null;
        }
    }

    /**
     * Extracts and decodes the first entry from a ZIP archive buffer.
     * Supports Store (method 0) and Deflate (method 8) compression.
     * Throws if the buffer is not a valid local-file ZIP entry.
     */
    private static async extractFirstZipEntry(buffer: Buffer): Promise<string> {
        // Full local file header signature: PK\x03\x04
        if (buffer.length < 30 || buffer[2] !== 0x03 || buffer[3] !== 0x04) {
            throw new Error('Invalid ZIP format: missing local file header');
        }

        // ZIP local file header layout (all values little-endian):
        //   offset  2 — version needed        (2 bytes)
        //   offset  6 — general purpose flags (2 bytes)
        //   offset  8 — compression method    (2 bytes)  0=store, 8=deflate
        //   offset 14 — CRC-32               (4 bytes)
        //   offset 18 — compressed size       (4 bytes)
        //   offset 22 — uncompressed size     (4 bytes)
        //   offset 26 — filename length       (2 bytes)
        //   offset 28 — extra field length    (2 bytes)
        //   offset 30 — filename + extra + data
        const compressionMethod = buffer.readUInt16LE(8);
        const compressedSize    = buffer.readUInt32LE(18);
        const filenameLen       = buffer.readUInt16LE(26);
        const extraLen          = buffer.readUInt16LE(28);
        const dataOffset        = 30 + filenameLen + extraLen;

        const compressedData = buffer.subarray(dataOffset, dataOffset + compressedSize);

        if (compressionMethod === 0) {
            // Store — data is uncompressed
            return compressedData.toString('latin1');
        }

        if (compressionMethod === 8) {
            // Deflate
            const decompressed = await inflateRawAsync(compressedData);
            return decompressed.toString('latin1');
        }

        throw new Error(`Unsupported ZIP compression method: ${compressionMethod}`);
    }

    /**
     * Creates a FIGFont from text content.
     */
    public static fromText(text: string | null): FIGFont | null {
        if (!text) {
            return null;
        }
        return FIGFont.fromLines(text.split('\n'));
    }

    /**
     * Creates a FIGFont from an array of lines.
     */
    public static fromLines(lines: string[] | null): FIGFont | null {
        if (!lines || lines.length === 0) {
            return null;
        }

        const font = new FIGFont();

        // Parse header
        const header = lines[0].trim();
        if (!header.startsWith("flf2a")) {
            throw new Error("Invalid FIGfont format");
        }

        const headerParts = header.split(' ');
        font.signature = headerParts[0];
        font.hardBlank = headerParts[0].substring(5, 6);
        font.height = parseInt(headerParts[1]);
        font.baseline = parseInt(headerParts[2]);
        font.maxLength = parseInt(headerParts[3]);
        font.oldLayout = parseInt(headerParts[4]);
        const commentLines = parseInt(headerParts[5]);
        
        if (headerParts.length > 6) {
            font.printDirection = parseInt(headerParts[6]);
        }
        if (headerParts.length > 7) {
            font.fullLayout = parseInt(headerParts[7]);
        }

        font.comments = lines.slice(1, 1 + commentLines).join('\n');

        if (isNaN(font.height) || isNaN(font.baseline) || isNaN(font.maxLength) || isNaN(font.oldLayout) || isNaN(commentLines)) {
            throw new Error("Invalid FIGfont header parameters");
        }

        if (lines.length < 1 + commentLines + font.height * (126 - 32 + 1)) {
            throw new Error("Not enough lines in FIGfont file for required characters");
        }

        // Skip header and comments
        let currentLine = 1 + commentLines;

        // Load required characters (ASCII 32-126)
        for (let charCode = 32; charCode <= 126; charCode++) {
            const endmark = FIGFont.detectEndmark(lines, currentLine);
            const endmarkRe = new RegExp(`[${endmark.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\n\\r]+$`);
            const charLines: string[] = new Array(font.height);
            for (let i = 0; i < font.height; i++) {
                charLines[i] = lines[currentLine + i]
                    .replace(endmarkRe, '')
                    // Special case for different hard blank character
                    .replace(/#$/, font.hardBlank === "#" ? "#" : "");
            }
            font.characters.set(String.fromCodePoint(charCode), charLines);
            currentLine += font.height;
        }

        // Continue reading additional characters if they exist
        while (currentLine + font.height <= lines.length) {
            const codeLine = lines[currentLine];
            if (!codeLine || !/^\d/.test(codeLine)) {
                break;
            }

            const codePoint = parseInt(codeLine.split(' ')[0]);
            currentLine++; // Move past the code point line

            const endmark = FIGFont.detectEndmark(lines, currentLine);
            const endmarkRe = new RegExp(`[${endmark.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\n\\r]+$`);
            const charLines: string[] = new Array(font.height);
            for (let i = 0; i < font.height; i++) {
                if (currentLine + i >= lines.length) break;
                charLines[i] = lines[currentLine + i].replace(endmarkRe, '');
            }
            font.characters.set(String.fromCodePoint(codePoint), charLines);
            currentLine += font.height;
        }

        // Parse the layout parameters
        FIGFont.parseLayoutParameters(font);

        return font;
    }

    /**
     * Loads the default FIGfont.
     */
    private static async loadDefaultFont(): Promise<FIGFont | null> {
        try {
            // Get the current module's directory
            const currentFileUrl = import.meta.url;
            const currentDir = dirname(fileURLToPath(currentFileUrl));
            
            // Construct path to the font file relative to the current module
            const fontPath = join(currentDir, 'assets', 'fonts', 'small.flf');
            
            // Read the font file directly using fs
            const fontContent = await fs.readFile(fontPath, 'utf-8');
            return FIGFont.fromText(fontContent);
        } catch (error) {
            console.error('Error loading default font:', error);
            return null;
        }
    }

    /**
     * Detects the endmark character used by a glyph block from the last character
     * of its first raw line (after stripping CR/LF).  Falls back to '@' if the
     * line is empty.
     */
    private static detectEndmark(lines: string[], blockStart: number): string {
        if (blockStart >= lines.length) return '@';
        const firstRaw = lines[blockStart].replace(/[\r\n]+$/, '');
        return firstRaw.length > 0 ? firstRaw[firstRaw.length - 1] : '@';
    }

    /**
     * Parses the layout parameters to determine smushing rules for the FIGfont.
     */
    private static parseLayoutParameters(font: FIGFont): void {
        // First, determine if we should use full_layout or old_layout
        let layoutMask: number;

        if (font.fullLayout > 0) {
            // Full layout is present
            layoutMask = font.fullLayout;

            // Check if horizontal smushing is enabled
            const horizontalSmushingEnabled = (layoutMask & 1) === 1;
            if (!horizontalSmushingEnabled) {
                font.smushingRules = SmushingRules.None;
                return;
            }

            // Extract just the rules part (bits 1-6)
            layoutMask = (layoutMask >> 1) & 0x3F;
        } else {
            // Use old layout
            layoutMask = font.oldLayout;

            // In old layout, -1 means no smushing
            if (layoutMask === -1) {
                font.smushingRules = SmushingRules.None;
                return;
            }

            // In old layout, 0 means kerning
            if (layoutMask === 0) {
                font.smushingRules = SmushingRules.None;
                return;
            }

            // For positive values, extract the rules
            if (layoutMask > 0) {
                // Convert old layout to new layout rules format
                layoutMask &= 0x3F;
            }
        }

        // Apply the final smushing rules
        font.smushingRules = layoutMask;
    }

    /**
     * Determines if the FIGfont has a specific smushing rule.
     */
    public hasSmushingRule(rule: SmushingRules): boolean {
        return (this.smushingRules & rule) === rule;
    }
}

File: FIGLet.Python/byteforge_figlet/fig_font.py
"""FIGFont parser — loads and exposes .flf font files."""

from __future__ import annotations

import importlib.resources
import io
import zipfile
from typing import Dict, List, Optional

from .smushing_rules import SmushingRules


class FIGFont:
    """Represents a FIGfont loaded from a .flf file."""

    # ------------------------------------------------------------------ #
    # Class-level default font cache
    # ------------------------------------------------------------------ #
    _default: Optional["FIGFont"] = None

    @classmethod
    @property
    def default(cls) -> "FIGFont":
        """Return the built-in 'small' FIGfont (loaded once and cached)."""
        if cls._default is None:
            cls._default = cls._load_default_font()
        if cls._default is None:
            raise RuntimeError("Default FIGfont could not be loaded")
        return cls._default

    # ------------------------------------------------------------------ #
    # Instance attributes
    # ------------------------------------------------------------------ #
    def __init__(self) -> None:
        self.signature: str = "flf2a"
        self.hard_blank: str = "#"
        self.height: int = 0
        self.baseline: int = 0
        self.max_length: int = 0
        self.old_layout: int = 0
        self.print_direction: int = 0
        self.full_layout: int = 0
        self.characters: Dict[int, List[str]] = {}
        self.smushing_rules: SmushingRules = SmushingRules.NONE
        self.comments: str = ""

    # ------------------------------------------------------------------ #
    # Factory methods
    # ------------------------------------------------------------------ #
    @classmethod
    def from_file(cls, path: Optional[str]) -> Optional["FIGFont"]:
        """Load a FIGFont from a file path (supports .flf and .zip)."""
        if not path:
            return None
        with open(path, "rb") as fh:
            return cls.from_stream(fh)

    @classmethod
    def from_stream(cls, stream: Optional[io.RawIOBase]) -> Optional["FIGFont"]:
        """Load a FIGFont from a binary stream (supports .flf inside .zip)."""
        if stream is None:
            return None
        data = stream.read()
        # Detect ZIP archive
        if data[:2] == b"PK":
            try:
                zf_obj = zipfile.ZipFile(io.BytesIO(data))
            except zipfile.BadZipFile:
                return None
            with zf_obj as zf:
                flf_names = [n for n in zf.namelist() if n.lower().endswith(".flf")]
                if not flf_names:
                    return None
                with zf.open(flf_names[0]) as entry:
                    data = entry.read()
        # Decode using latin-1 (FIGlet standard)
        try:
            text = data.decode("latin-1")
        except UnicodeDecodeError:
            text = data.decode("utf-8", errors="replace")
        return cls.from_lines(text.splitlines())

    @classmethod
    def from_text(cls, text: Optional[str]) -> Optional["FIGFont"]:
        """Load a FIGFont from a string containing the full .flf content."""
        if not text:
            return None
        return cls.from_lines(text.splitlines())

    @classmethod
    def from_lines(cls, lines: Optional[List[str]]) -> Optional["FIGFont"]:
        """Load a FIGFont from a list of lines."""
        if not lines:
            return None

        font = cls()

        # --- Parse header ---
        header = lines[0].rstrip()
        if not header.startswith("flf2a"):
            raise ValueError("Invalid FIGfont format: header must start with 'flf2a'")

        try:
            parts = header.split(" ")
            font.signature = parts[0]
            font.hard_blank = parts[0][5] if len(parts[0]) > 5 else "#"
            font.height = int(parts[1])
            font.baseline = int(parts[2])
            font.max_length = int(parts[3])
            font.old_layout = int(parts[4])
            comment_lines = int(parts[5])
            if len(parts) > 6:
                try:
                    font.print_direction = int(parts[6])
                except ValueError:
                    pass
            if len(parts) > 7:
                try:
                    font.full_layout = int(parts[7])
                except ValueError:
                    pass
        except (IndexError, ValueError) as exc:
            raise ValueError(f"Error parsing FIGfont header: {exc}") from exc

        # --- Collect comments ---
        font.comments = "\n".join(lines[1 : 1 + comment_lines])

        current_line = 1 + comment_lines

        # --- Load required characters ASCII 32–126 ---
        for code_point in range(32, 127):
            char_lines = cls._read_char_lines(lines, font, current_line)
            font.characters[code_point] = char_lines
            current_line += font.height

        # --- Load optional extra characters ---
        while current_line < len(lines):
            code_line = lines[current_line].strip()
            if not code_line or not (code_line[0].isdigit() or code_line.startswith("-")):
                break
            # The code point is the first token; optional comment follows
            token = code_line.split(None, 1)[0]
            try:
                code_point = cls._parse_int(token)
            except ValueError:
                break
            current_line += 1
            if current_line + font.height > len(lines):
                break
            char_lines = cls._read_char_lines(lines, font, current_line)
            font.characters[code_point] = char_lines
            current_line += font.height

        # --- Derive smushing rules ---
        cls._parse_layout_parameters(font)

        return font

    # ------------------------------------------------------------------ #
    # Helpers
    # ------------------------------------------------------------------ #
    @staticmethod
    def _read_char_lines(lines: List[str], font: "FIGFont", start: int) -> List[str]:
        result: List[str] = []
        for i in range(font.height):
            raw = lines[start + i] if start + i < len(lines) else ""
            # Strip trailing @ markers and newline characters
            stripped = raw.rstrip("\r\n").rstrip("@")
            # Workaround: some fonts use '#' as EOL even when hardblank != '#'
            if stripped.endswith("#") and font.hard_blank != "#":
                stripped = stripped.rstrip("#")
            result.append(stripped)
        return result

    @staticmethod
    def _parse_int(text: str) -> int:
        text = text.strip()
        if not text:
            return 0
        if text.startswith("-"):
            return -FIGFont._parse_int(text[1:])
        if text.lower().startswith("0x"):
            return int(text[2:], 16)
        if text.lower().startswith("0b"):
            return int(text[2:], 2)
        if len(text) > 1 and text.startswith("0"):
            return int(text, 8)
        return int(text)

    @staticmethod
    def _parse_layout_parameters(font: "FIGFont") -> None:
        if font.full_layout > 0:
            layout_mask = font.full_layout
            horizontal_smushing_enabled = (layout_mask & 1) == 1
            if not horizontal_smushing_enabled:
                font.smushing_rules = SmushingRules.NONE
                return
            layout_mask = (layout_mask >> 1) & 0x3F
        else:
            layout_mask = font.old_layout
            if layout_mask <= 0:
                font.smushing_rules = SmushingRules.NONE
                return
            layout_mask &= 0x3F

        font.smushing_rules = SmushingRules(layout_mask)

    @classmethod
    def _load_default_font(cls) -> Optional["FIGFont"]:
        try:
            pkg = importlib.resources.files("byteforge_figlet") / "fonts" / "small.flf"
            data = pkg.read_bytes()
            return cls.from_stream(io.BytesIO(data))
        except Exception:
            return None

    # ------------------------------------------------------------------ #
    # Public helpers
    # ------------------------------------------------------------------ #
    def has_smushing_rule(self, rule: SmushingRules) -> bool:
        """Return True if the font defines the given smushing rule."""
        return self.smushing_rules.has_rule(rule)
